std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
-                        prevToken(&tok_it, tree);
-                    }
-                }
-                continue;
-            },
-            State.ParamDeclEnd => |ctx| {
-                if (eatToken(&tok_it, tree, Token.Id.Ellipsis3)) |ellipsis3| {
-                    ctx.param_decl.var_args_token = ellipsis3;
-
-                    switch (expectCommaOrEnd(&tok_it, tree, Token.Id.RParen)) {
-                        ExpectCommaOrEndResult.end_token => |t| {
-                            if (t == null) {
-                                stack.append(State{ .ExpectToken = Token.Id.RParen }) catch unreachable;
-                                continue;
-                            }
-                            continue;
-                        },
-                        ExpectCommaOrEndResult.parse_error => |e| {
-                            try tree.errors.push(e);
-                            return tree;
-                        },
-                    }
-                }
+    const type_expr = blk: {
+        if (eatToken(it, .Colon)) |_| {
+            break :blk try expectNode(arena, it, tree, parseTypeExpr, AstError{
+                .ExpectedTypeExpr = AstError.ExpectedTypeExpr{ .token = it.index },
+            });
+        } else break :blk null;
+    };
 
-                try stack.append(State{ .ParamDeclComma = ctx.fn_proto });
-                try stack.append(State{ .TypeExprBegin = OptionalCtx{ .Required = &ctx.param_decl.type_node } });
-                continue;
-            },
-            State.ParamDeclComma => |fn_proto| {
-                switch (expectCommaOrEnd(&tok_it, tree, Token.Id.RParen)) {
-                    ExpectCommaOrEndResult.end_token => |t| {
-                        if (t == null) {
-                            stack.append(State{ .ParamDecl = fn_proto }) catch unreachable;
-                        }
-                        continue;
-                    },
-                    ExpectCommaOrEndResult.parse_error => |e| {
-                        try tree.errors.push(e);
-                        return tree;
-                    },
-                }
-            },
+    // TODO: supply default value to struct field when ast.Node.StructField supports it
+    const default_value = blk: {
+        if (eatToken(it, .Equal)) |_| {
+            break :blk try expectNode(arena, it, tree, parseExpr, AstError{
+                .ExpectedExpr = AstError.ExpectedExpr{ .token = it.index },
+            });
+        } else break :blk null;
+    };
 
-            State.MaybeLabeledExpression => |ctx| {
-                if (eatToken(&tok_it, tree, Token.Id.Colon)) |_| {
-                    stack.append(State{
-                        .LabeledExpression = LabelCtx{
-                            .label = ctx.label,
-                            .opt_ctx = ctx.opt_ctx,
-                        },
-                    }) catch unreachable;
-                    continue;
-                }
+    switch (kind) {
```

Comment:
I'm gonna go ahead and implement your suggestion, partly for the cleanup and partly because it actually conforms better to the grammar. Fortunately (?), none of these are implemented in `ir.zig` yet, so the impact on outside code is negligible.